core/src/main/java/ch/cyberduck/core/AbstractProtocol.java

```diff
@@ -39,11 +39,137 @@
 import java.util.Set;
 import java.util.stream.Collectors;
 
+import static ch.cyberduck.core.Profile.*;
+
 public abstract class AbstractProtocol implements Protocol {
 
     @Override
     public <T> T serialize(final Serializer<T> dict) {
-        return null;
+        if(this.getOAuthClientId() != null) {
+            dict.setStringForKey(this.getOAuthClientId(), OAUTH_CLIENT_ID_KEY);
+        }
+        if(this.getOAuthClientSecret() != null) {
+            dict.setStringForKey(this.getOAuthClientSecret(), OAUTH_CLIENT_SECRET_KEY);
+        }
+        if(this.getOAuthTokenUrl() != null) {
+            dict.setStringForKey(this.getOAuthTokenUrl(), OAUTH_TOKEN_URL_KEY);
+        }
+        if(this.getOAuthRedirectUrl() != null) {
+            dict.setStringForKey(this.getOAuthRedirectUrl(), OAUTH_REDIRECT_URL_KEY);
+        }
+        if(this.getOAuthAuthorizationUrl() != null) {
+            dict.setStringForKey(this.getOAuthAuthorizationUrl(), OAUTH_AUTHORIZATION_URL_KEY);
+        }
+        if(!this.isOAuthPKCE()) {
+            dict.setBooleanForKey(this.isOAuthPKCE(), OAUTH_PKCE_KEY);
+        }
+        if(!this.getOAuthScopes().isEmpty()) {
+            dict.setStringListForKey(this.getOAuthScopes(), SCOPES_KEY);
+        }
+        if(this.getSTSEndpoint() != null) {
+            dict.setStringForKey(this.getSTSEndpoint(), STS_ENDPOINT_KEY);
+        }
+        if(this.disk() != null) {
+            dict.setStringForKey(this.disk(), DISK_KEY);
+        }
+        if(this.icon() != null) {
+            dict.setStringForKey(this.icon(), ICON_KEY);
+        }
+        if(this.getIdentifier() != null) {
+            dict.setStringForKey(this.getIdentifier(), PROTOCOL_KEY);
```

Comment:
@dkocher bad code smell, we implement `AbstractHostProtocol.serialize` but want to (mis-)use at as a builder for profile serialization?

Reply:
I do not understand the concern.

Reply:
`VENDOR_KEY`/`PROTOCOL_KEY` is defined in `Profile` (as it should), as `AbstractProtocol` is not a sub-class of `Profile`, it should have no notion of `VENDOR_KEY`/`PROTOCOL_KEY`. It works but is confusing imho 😕